src/app/clusters/basic-information/BasicInformationCluster.cpp

```diff
@@ -247,175 +254,194 @@ inline CHIP_ERROR ReadProductAppearance(DeviceInstanceInfoProvider * deviceInfoP
+{
+#define OPTIONAL_ATTR(name) { mEnabledOptionalAttributes.Has(OptionalBasicInformationAttributes::k##name), name::kMetadataEntry }
+
+    AttributeListBuilder::OptionalAttributeEntry optionalAttributes[] = {
+        { !mEnabledOptionalAttributes.Has(OptionalBasicInformationAttributes::kDisableMandatoryUniqueIDOnPurpose),
+          UniqueID::kMetadataEntry },
+        OPTIONAL_ATTR(ManufacturingDate),
+        OPTIONAL_ATTR(PartNumber),
+        OPTIONAL_ATTR(ProductURL),
+        OPTIONAL_ATTR(ProductLabel),
+        OPTIONAL_ATTR(SerialNumber),
+        OPTIONAL_ATTR(LocalConfigDisabled),
+        OPTIONAL_ATTR(Reachable),
+        OPTIONAL_ATTR(ProductAppearance),
+    };
+
+    AttributeListBuilder listBuilder(builder);
+
+    return listBuilder.Append(Span(kMandatoryAttributes), Span(optionalAttributes));
 }
 
-class PlatformMgrDelegate : public DeviceLayer::PlatformManagerDelegate
+CHIP_ERROR BasicInformationCluster::Startup(ServerClusterContext & context)
 {
-    void OnStartUp(uint32_t softwareVersion) override
+    ReturnErrorOnFailure(DefaultServerCluster::Startup(context));
+    if (PlatformMgr().GetDelegate() == nullptr)
     {
-        MATTER_TRACE_INSTANT("OnStartUp", "BasicInfo");
-        // The StartUp event SHALL be emitted by a Node after completing a boot or reboot process
-        ChipLogDetail(Zcl, "Emitting StartUp event");
-
-        for (auto endpoint : EnabledEndpointsWithServerCluster(BasicInformation::Id))
-        {
-            // If Basic cluster is implemented on this endpoint
-            Events::StartUp::Type event{ softwareVersion };
-            EventNumber eventNumber;
-
-            CHIP_ERROR err = LogEvent(event, endpoint, eventNumber);
-            if (CHIP_NO_ERROR != err)
-            {
-                ChipLogError(Zcl, "Failed to emit StartUp event: %" CHIP_ERROR_FORMAT, err.Format());
-            }
-        }
+        PlatformMgr().SetDelegate(this);
     }
 
-    void OnShutDown() override
-    {
-        MATTER_TRACE_INSTANT("OnShutDown", "BasicInfo");
-        // The ShutDown event SHOULD be emitted on a best-effort basis by a Node prior to any orderly shutdown sequence.
-        ChipLogDetail(Zcl, "Emitting ShutDown event");
-
-        for (auto endpoint : EnabledEndpointsWithServerCluster(BasicInformation::Id))
-        {
-            // If Basic cluster is implemented on this endpoint
-            Events::ShutDown::Type event;
-            EventNumber eventNumber;
-
-            CHIP_ERROR err = LogEvent(event, endpoint, eventNumber);
-            if (CHIP_NO_ERROR != err)
-            {
-                ChipLogError(Zcl, "Failed to emit ShutDown event: %" CHIP_ERROR_FORMAT, err.Format());
-            }
-        }
-
-        // Flush the events to increase chances that they get sent before the shutdown
-        InteractionModelEngine::GetInstance()->GetReportingEngine().ScheduleUrgentEventDeliverySync();
```

Comment:
This ScheduleUrgentEventDeliverySync() call is critical.  I don't see anything replacing it, unless I am seriously missing something.  Changes requested for this part; not having this breaks shutdown events completely, so we should either do this or stop generating them, because no one will ever see them without this call.

Reply:
we have to figure out how to set this in general - I do not want a binding to the singleton, so this needs some urgency bit into our markdirty.